python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,300 @@
+import os
+from concurrent import futures
+from typing import List, NamedTuple, Tuple, Type, cast
+
+import numba
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+import pyarrow as pa
+from scipy import sparse
+
+import somacore.data as scd
+from somacore.query import eager_iter
+
+
+class CSRAccumulatorFinalResult(NamedTuple):
+    """
+    Return type for the CSRAccumulator.finalize method.
+    Contains a sparse CSR consituent elements
+    """
+
+    data: npt.NDArray[np.number]
+    indptr: npt.NDArray[np.integer]
+    indices: npt.NDArray[np.integer]
+    shape: Tuple[int, int]
+
+
+class CSRAccumulator:
+    """
+    Fast accumulator of a CSR, based upon COO input.
+    """
+
+    def __init__(
+        self,
+        obs_joinids: npt.NDArray[np.int64],
+        var_joinids: npt.NDArray[np.int64],
+        pool: futures.Executor,
+    ):
+        self.obs_joinids = obs_joinids
+        self.var_joinids = var_joinids
+        self.pool = pool
+
+        self.shape: Tuple[int, int] = (len(self.obs_joinids), len(self.var_joinids))
+        self.obs_indexer: pd.Index = pd.Index(self.obs_joinids)
+        self.var_indexer: pd.Index = pd.Index(self.var_joinids)
+        self.row_length: npt.NDArray[np.int64] = np.zeros(
+            (self.shape[0],), dtype=_select_dtype(self.shape[1])
+        )
+
+        # COO accumulated chunks, stored as list of triples (data, row_ind, col_ind)
+        self.coo_chunks: List[
+            Tuple[
```

Comment:
potential thought: make this a CSRAccumulatorFinalResult, and just don't use the `shape` field?

Reply:
they are quite different.  One is COO, the other is CSR.  I would find this change very confusing.

Reply:
If you are asking for this to be put into a named tuple, I can add that - but it will need to be a different type than CSRAccumulatorFinalResult. I may also need to convert to an actual tuple for numba, as I believe numba.List only supports native tuples.